### PR TITLE
migrate to indy okhttp plugin

### LINKED ISSUE #1337: Migrate plugins to indy plugins
A meta-issue that tracks the migration

- [x] apm-apache-httpclient-plugin https://github.com/elastic/apm-agent-java/pull/1349
- [x] apm-api-plugin https://github.com/elastic/apm-agent-java/pull/1411
- [x] apm-asynchttpclient-plugin elastic/apm-agent-java#1351
- [x] apm-bootdelegation-plugin N/A - might be removed once all plugins have been migrated to indy plugins
- [x] apm-dubbo-plugin https://github.com/elastic/apm-agent-java/pull/2087
- [x] apm-error-logging-plugin https://github.com/elastic/apm-agent-java/pull/1362
- [x] apm-es-restclient-plugin #2088
- [x] apm-grails-plugin https://github.com/elastic/apm-agent-java/pull/1371
- [x] apm-grpc https://github.com/elastic/apm-agent-java/issues/1358
- [x] apm-hibernate-search-plugin https://github.com/elastic/apm-agent-java/pull/1404
- [x] apm-httpclient11-plugin
- [x] apm-httpclient-core n/a
- [x] apm-java-concurrent-plugin
- [x] apm-jaxrs-plugin https://github.com/elastic/apm-agent-java/pull/1374
- [x] apm-jaxws-plugin elastic/apm-agent-java#1383
- [x] apm-jdbc-plugin
- [x] apm-jms-plugin https://github.com/elastic/apm-agent-java/pull/1394
- [x] apm-jmx-plugin N/A
- [x] apm-jsf-plugin https://github.com/elastic/apm-agent-java/pull/1405
- [x] apm-kafka-plugin #2095 
- [x] apm-log-correlation-plugin N/A
- [x] apm-log-shipper-plugin N/A
- [x] apm-micrometer-plugin
- [x] apm-mongoclient-plugin https://github.com/elastic/apm-agent-java/pull/1369
- [x] apm-okhttp-plugin https://github.com/elastic/apm-agent-java/pull/1368
- [x] apm-opentracing-plugin #1410
- [x] apm-process-plugin
- [x] apm-profiling-plugin N/A
- [x] apm-quartz-job-plugin https://github.com/elastic/apm-agent-java/pull/1366
- [x] apm-redis-plugin #2090
- [x] apm-scala-concurrent-plugin https://github.com/elastic/apm-agent-java/pull/1367
- [x] apm-scheduled-annotation-plugin https://github.com/elastic/apm-agent-java/pull/1364
- [x] apm-servlet-plugin
- [x] apm-spring-resttemplate-plugin https://github.com/elastic/apm-agent-java/pull/1363
- [x] apm-spring-webmvc-plugin https://github.com/elastic/apm-agent-java/pull/1365
- [x] apm-urlconnection-plugin
- [x] trace_methods instrumentation #2094 

Prerequisite for elastic/apm-agent-java#1338 and elastic/apm-agent-java#1339

### PR DESCRIPTION
**Pull request by kananindzya**
refer #1337

### CODE REVIEW

File: apm-agent-plugins/apm-okhttp-plugin/src/main/java/co/elastic/apm/agent/okhttp/OkHttp3ClientInstrumentation.java [hunk 1]
```diff
@@ -40,62 +36,52 @@
...
     @Override
     public Class<?> getAdviceClass() {
         return OkHttpClient3ExecuteAdvice.class;
     }
 
-    @VisibleForAdvice
     public static class OkHttpClient3ExecuteAdvice {
 
-        @VisibleForAdvice
-        public final static GlobalThreadLocal<Span> spanTls = GlobalThreadLocal.get(OkHttpClient3ExecuteAdvice.class, "spanTls");
```
Comment (felixbarny) [inline apm-agent-plugins/apm-okhttp-plugin/src/main/java/co/elastic/apm/agent/okhttp/OkHttp3ClientInstrumentation.java:63]: It's probably fine either way, so just out of curiosity, why did you change from `GlobalThreadLocal<Span>` to returning the span in the `Object[]` return value?

**Review by felixbarny — APPROVED** [review]

File: apm-agent-plugins/apm-okhttp-plugin/src/main/java/co/elastic/apm/agent/okhttp/OkHttp3ClientInstrumentation.java (hunk 1 above)
Comment (kananindzya) [reply apm-agent-plugins/apm-okhttp-plugin/src/main/java/co/elastic/apm/agent/okhttp/OkHttp3ClientInstrumentation.java:63]: Hah, good question :) I don't know why.
What is the difference between using GlobalThreadLocal and returning an array of objects? And when to use them?

Comment (felixbarny) [reply apm-agent-plugins/apm-okhttp-plugin/src/main/java/co/elastic/apm/agent/okhttp/OkHttp3ClientInstrumentation.java:63]: Also a good question :)
I guess in terms of allocations, the `Obect[]` approach might even be better, as the JIT compiler should be able to do some escape analysis and replace it with a stack allocation. Also, it seems a bit safer maybe as there are fewer things that can go wrong, such as forgetting to clear the ThreadLocal.
A downside is that it requires more casting and things can go wrong when accessing the wrong index in the array.

But overall, the `Object[]` approach is probably preferable in most situations.

**Review by SylvainJuge — APPROVED** [review]
